src/util/apparmor.c

```diff
@@ -93,3 +105,446 @@ int bus_apparmor_dbus_supported(bool *supportedp) {
+                 * names and match each. Note that the snapshot must contain
+                 * queued names as well, since the policy matches on it.
+                 */
+                for (i = 0; i < nameset->snapshot->n_names; ++i) {
+                        r = apparmor_message_query_name(check_send, security_label, bustype,
+                                                        receiver_context,
+                                                        nameset->snapshot->names[i]->name,
+                                                        path, interface, method, allow, &audit_tmp);
+                        if (r < 0)
+                                return r;
+                        if (!allow)
+                                return r;
+                        if (audit_tmp)
+                                *audit = 1;
+                }
+        } else if (nameset->type != NAME_SET_TYPE_EMPTY) {
+                c_assert(0);
+                r = -EINVAL;
+        }
+
+        return r;
+}
+
+/**
+ * bus_apparmor_check_own() - check if the given transaction is allowed
+ * @registry:           AppArmor registry to operate on
+ * @context:            security context requesting the name
+ * @name:               name to be owned
+ *
+ * Check if the given owner context is allowed to own the given name.
+ *
+ * The contexts are pinned when peers connect, and as such could in principle
+ * become invalid in case a new policy is loaded that does not know the
+ * old labels. In this case we treat this as if the ownership request was
+ * denied.
+ *
+ * Return: 0 if the ownership is allowed, BUS_APPARMOR_E_DENIED if it is not,
+ *         or a negative error code on failure.
+ */
+int bus_apparmor_check_own(struct BusAppArmorRegistry *registry,
+                           const char *owner_context,
+                           const char *name) {
+        _c_cleanup_(c_freep) char *condup = NULL;
+        char *security_label, *security_mode;
+        _c_cleanup_(c_freep) char *qstr = NULL;
+        int r;
+        /* the AppArmor API uses pointers to int for pointers to boolean */
+        int allow = false, audit = true;
+
+        if (!is_apparmor_enabled(registry))
+                return 0;
+
+        condup = strdup(owner_context);
+        if (!condup)
+                return error_origin(-ENOMEM);
+
+        security_label = aa_splitcon(condup, &security_mode);
+
+        r = build_service_query(&qstr, security_label, registry->bustype, name);
+        if (r < 0)
+                return error_origin(r);
+
+        r = aa_query_label(AA_DBUS_BIND, qstr, r, &allow, &audit);
+        if (r < 0)
+                return error_origin(-errno);
+
+        if (audit)
+                bus_apparmor_log("apparmor=\"%s\" operation=\"dbus_bind\" bus=\"%s\" name=\"%s\"", allowstr(allow), registry->bustype, name);
+
+        if (is_complain(security_mode))
+                allow = 1;
+
```

Comment:
Should this `allow = 1` block be moved above the `bus_apparmor_log()` call, so that the `allowstr(allow)` generates an output that matches the decision?

Reply:
Sounds good!

Reply:
Thanks!